src/category_theory/monad/adjunction.lean

```diff
@@ -0,0 +1,133 @@
+
+end monad
+
+class reflective (R : D ⥤ C) extends is_right_adjoint R, full R, faithful R.
+
+instance μ_iso_of_reflective [reflective R] : is_iso (μ_ ((left_adjoint R) ⋙ R)) :=
+by { dsimp [adjunction.monad], apply_instance }
+
+class monadic_right_adjoint (R : D ⥤ C) extends is_right_adjoint R :=
+(eqv : is_equivalence (monad.comparison R))
+
+attribute [instance] monadic_right_adjoint.eqv
+
+-- PROJECT prove Beck's monadicity theorem, e.g. from Section 5.5 of Riehl
+
+namespace reflective
+
+lemma comparison_ess_surj_aux [reflective R] (X : monad.algebra ((left_adjoint R) ⋙ R)) :
+  ((is_right_adjoint.adj R).unit).app (R.obj ((left_adjoint R).obj (X.A))) = R.map ((left_adjoint R).map ((is_right_adjoint.adj R).unit.app X.A)) :=
+begin
+ -- both are left inverses to μ_X.
+ apply (cancel_mono ((μ_ ((left_adjoint R) ⋙ R)).app _)).1,
+ { dsimp, erw [adjunction.right_triangle_components, ←R.map_comp], simp, },
+ { apply is_iso.mono_of_iso _,
+   apply nat_iso.is_iso_app_of_is_iso }
+end
+
+instance [reflective R] (X : monad.algebra ((left_adjoint R) ⋙ R)) :
+  is_iso ((is_right_adjoint.adj R).unit.app X.A) :=
+let L := left_adjoint R in
+let h := (is_right_adjoint.adj R) in
+{ inv := X.a,
+  hom_inv_id' := X.unit,
+  inv_hom_id' :=
+  begin
+    dsimp,
+    erw [h.unit.naturality, comparison_ess_surj_aux,
+          ←R.map_comp, ←L.map_comp, X.unit, L.map_id, R.map_id],
+    refl
+  end }
+
+instance comparison_ess_surj [reflective R]: ess_surj (monad.comparison R) :=
+let L := left_adjoint R in
+let h := (is_right_adjoint.adj R) in
+{ obj_preimage := λ X, L.obj X.A,
+  iso' := λ X,
+  { hom :=
+    { f := (as_iso (h.unit.app X.A)).inv,
+      h' :=
+      begin
+        dsimp,
+        apply (cancel_epi (R.map (L.map ((h.unit).app (X.A))))).1,
+        rw [is_iso.hom_inv_id_assoc, ←category.assoc, ←R.map_comp,adjunction.left_triangle_components],
+        erw [functor.map_id, category.id_comp],
+        apply (cancel_epi ((h.unit).app (X.A))).1,
+        rw is_iso.hom_inv_id,
+        exact X.unit,
+      end },
+    inv :=
+    { f := (as_iso (h.unit.app X.A)).hom,
+      h' :=
+      begin
+        dsimp,
+        erw [←R.map_comp, adjunction.left_triangle_components, R.map_id],
+        apply (cancel_epi ((h.unit).app (X.A))).1,
+        conv { to_rhs, erw [←category.assoc, X.unit] },
+        erw [comp_id, id_comp],
+      end },
+    hom_inv_id' := by { ext, exact (as_iso (h.unit.app X.A)).inv_hom_id, },
+    inv_hom_id' := by { ext, exact (as_iso (h.unit.app X.A)).hom_inv_id, }, } }
+
+instance comparison_full [full R] [is_right_adjoint R] : full (monad.comparison R) :=
+{ preimage := λ X Y f, R.preimage f.f }
+instance comparison_faithful [faithful R] [is_right_adjoint R] : faithful (monad.comparison R) :=
+{ injectivity' := λ X Y f g w, by { have w' := (congr_arg monad.algebra.hom.f w), exact R.injectivity w' } }
+
+end reflective
+
+-- Proposition 5.3.3 of Riehl
+instance monadic_of_reflective [reflective R] : monadic_right_adjoint R :=
```

Comment:
The converse is of course also true... every `monadic_right_adjoint` is `reflective`. One of  the two can be an instance, the other cannot. I'm inclined to say that the converse is "obvious" and therefore should be an instance, whereas the current direction is a "theorem" which people should invoke manually.